### PR TITLE
Simplify diff calculation for remote routing table

### PR DESCRIPTION
**Pull request by Bukhtawar**
### Description
Simplifies diff calculation for remote routing table and makes it compatible with Diff<RoutingTable>. This change make it simpler to switch between complete diff and incremental diffs.

Some changes in the PR might not be backward compatible with existing Public APIs and would be addressed as a follow up PR 

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java [hunk 1]
```diff
@@ -513,14 +513,14 @@ UploadedMetadataResults writeMetadataInParallel(
...
-        Map<String, Diff<IndexRoutingTable>> indexRoutingTableDiff
+        RoutingTable previousRoutingTable
     ) throws IOException {
         assert Objects.nonNull(indexMetadataUploadListeners) : "indexMetadataUploadListeners can not be null";
         int totalUploadTasks = indexToUpload.size() + indexMetadataUploadListeners.size() + customToUpload.size()
             + (uploadCoordinationMetadata ? 1 : 0) + (uploadSettingsMetadata ? 1 : 0) + (uploadTemplateMetadata ? 1 : 0)
             + (uploadDiscoveryNodes ? 1 : 0) + (uploadClusterBlock ? 1 : 0) + (uploadTransientSettingMetadata ? 1 : 0)
             + clusterStateCustomToUpload.size() + (uploadHashesOfConsistentSettings ? 1 : 0) + indicesRoutingToUpload.size()
-            + (indexRoutingTableDiff != null && !indexRoutingTableDiff.isEmpty() ? 1 : 0);
+            + (previousRoutingTable != null && !previousRoutingTable.indicesRouting().isEmpty() ? 1 : 0);
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:523]: Will this break if there is no diff but listener is waiting for the diff task?

File: server/src/main/java/org/opensearch/gateway/remote/ClusterStateDiffManifest.java [hunk 2]
```diff
@@ -344,12 +313,6 @@ public static ClusterStateDiffManifest fromXContent(XContentParser parser, long
                         currentFieldName = parser.currentName();
                         parser.nextToken();
                         switch (currentFieldName) {
-                            case UPSERTS_FIELD:
-                                builder.indicesRoutingUpdated(convertListToString(parser.listOrderedMap()));
-                                break;
-                            case DELETES_FIELD:
-                                builder.indicesRoutingDeleted(convertListToString(parser.listOrderedMap()));
-                                break;
                             case DIFF_FIELD:
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/gateway/remote/ClusterStateDiffManifest.java:316]: Do we need to bump up the CODEC version for BWC?

Comment (shailendra0811) [reply server/src/main/java/org/opensearch/gateway/remote/ClusterStateDiffManifest.java]: Made changes to increment CODEC version.

**Review by shwetathareja — APPROVED** [review]

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java [hunk 3]
```diff
@@ -1398,9 +1412,11 @@ public ClusterState getClusterStateUsingDiff(ClusterMetadataManifest manifest, C
         }
 
         List<UploadedIndexMetadata> updatedIndexRouting = new ArrayList<>();
-        updatedIndexRouting.addAll(
-            remoteRoutingTableService.getUpdatedIndexRoutingTableMetadata(diff.getIndicesRoutingUpdated(), manifest.getIndicesRouting())
-        );
+        if (manifest.getCodecVersion() == CODEC_V2 || manifest.getCodecVersion() == CODEC_V3) {
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:1415]: Should we just check Codec < V4?

**Pull request merged**